libmamba/include/mamba/util/flat_bool_expr_tree.hpp

```diff
@@ -0,0 +1,738 @@
+
+        using branch_type = Branch;
+        using leaf_type = Leaf;
+        struct BranchNode
+        {
+            branch_type data;
+            std::size_t left_child = 0;
+            std::size_t right_child = 0;
+        };
+        using branch_node = BranchNode;
+        using leaf_node = leaf_type;
+        using node_type = std::variant<branch_node, leaf_node>;
+        using node_list = std::vector<node_type>;
+        using size_type = typename node_list::size_type;
+        using idx_type = size_type;
+
+        [[nodiscard]] auto size() const -> size_type;
+        [[nodiscard]] auto empty() const -> bool;
+
+        void clear();
+        void reserve(size_type size);
+
+        auto add_leaf(const leaf_type& leaf) -> idx_type;
+        auto add_leaf(leaf_type&& leaf) -> idx_type;
+        auto add_branch(const branch_type& branch, idx_type left_child, idx_type right_child)
+            -> idx_type;
+        auto add_branch(branch_type&& branch, idx_type left_child, idx_type right_child) -> idx_type;
+
+        [[nodiscard]] auto node(idx_type idx) const -> const node_type&;
+        [[nodiscard]] auto node(idx_type idx) -> node_type&;
+        [[nodiscard]] auto is_branch(idx_type idx) const -> bool;
+        [[nodiscard]] auto is_leaf(idx_type idx) const -> bool;
+        [[nodiscard]] auto leaf(idx_type idx) const -> const leaf_type&;
+        [[nodiscard]] auto leaf(idx_type idx) -> leaf_type&;
+        [[nodiscard]] auto branch(idx_type idx) const -> const branch_type&;
+        [[nodiscard]] auto branch(idx_type idx) -> branch_type&;
+        [[nodiscard]] auto left(idx_type idx) const -> idx_type;
+        [[nodiscard]] auto right(idx_type idx) const -> idx_type;
+        [[nodiscard]] auto root() const -> idx_type;
+
+    private:
+
+        node_list m_nodes;
+        idx_type m_root = 0;
+
+        template <typename L>
+        auto add_leaf_impl(L&& leaf) -> idx_type;
+        template <typename B>
+        auto add_branch_impl(B&& branch, idx_type left_child, idx_type right_child) -> idx_type;
+    };
+
+    /**
+     * A parser for postfix expressions.
+     *
+     * The parser creates an expression tree and validate that the expression being pushed
+     * is a valid postfix expression.
+     * For example, for the expression ``a + b * c`` on might push ``a b c * +``
+     * or ``b c * a +``.
+     */
+    template <typename Variable, typename Operator>
+    class PostfixParser
+    {
+    public:
+
+        using operator_type = Operator;
+        using variable_type = Variable;
+        using tree_type = flat_binary_tree<operator_type, variable_type>;
+
+        void push_variable(const variable_type& var);
+        void push_variable(variable_type&& var);
+        void push_operator(const operator_type& op);
+        void push_operator(operator_type&& op);
+        void finalize();
+
+        [[nodiscard]] auto tree() const& -> const tree_type&;
+        [[nodiscard]] auto tree() && -> tree_type&&;
```

Comment:
Im not completely sure if it's ok to keep that same name here, as it could lead to unexpectedly extract the tree from theparser and making it unusable silently. Did you consider renaming this function something like `extract_tree()` to clarify?

Reply:
IMHO this is the same style as rvalue overloads as `std::optional<T>::value` `std::get`...

Extracting the tree is really the point of the parser anyways.